0004-remove-spandata.md

```diff
@@ -0,0 +1,43 @@
+# Remove SpanData
+
+Remove and replace SpanData by adding span start and finish options.
+
+## Motivation
+
+SpanData represents an immutable span object, creating a fairly large API for all of the fields (12 to be exact). It exposes what feels like an SDK concern and implementation detail to the API surface. As a user, this is another API I need to learn how to use. As an implementer, it’s a new data type that needs to be supported. The primary motivation for removing SpanData revolves around the desire to reduce the size of the tracing API.
+
+## Explanation
+
+SpanData's primary use case comes from the need to construct and report out of band spans, meaning that you're creating "custom" spans for an operation you don't own. A good example of this is a program that takes in structured logs that contain correlation IDs and a duration (e.g. from splunk) and [converts them](https://github.com/lightstep/splunktospan/blob/master/splunktospan/span.py#L43) to spans for your tracing system. [Another example](https://github.com/lightstep/haproxy_log2span/blob/master/lib/lib.go#L292) is running a sidecar on an HAProxy machine, tailing the request logs, and creating spans. SpanData supports the out of band reporting case, whereas you can’t with the current Span API as you cannot set the start and end timestamp, or add any tags at span creation that the sampler might need.
+
+I'd like to propose getting rid of SpanData and `tracer.recordSpanData()` and replacing it by allowing `tracer.startSpan()` and `span.finish()` to accept start and end timestamp options. This reduces the API surface, consolidating on a single span type. Options would meet the requirements for out of band reporting.
+
+## Internal details
+
+`startSpan()` would change so you can include options such as `withStartTimestamp()`, `withTags()`, `withResource()`, `withEvents()`, `withLogs()`, etc. and for `finish()` you could have `withEndTimestamp()`, `withEvents()`, `withLogs()`, etc. The exact implementation would be language specific, so some would use an options pattern with function overloading or variadic parameters, or a span builder pattern.
```

Comment:
Adding events/links/etc. has been supported by span APIs already (e.g. [`span.addEvent`](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/tracing-api.md#addevent)), what's the motivation to duplicate such functionality in `startSpan()`?

Reply:
There's this hidden dependency on sampling that I've heard come up, where you might want to look at the tags or start+end timestamps of SpanData. If sampling occurs in `startSpan()`, you'd want to be able to provide information in bulk upfront for the sampler. I should probably clarify that the RFC specifically wants to add start and end timestamp options, the rest is optional and should be discussed.

Reply:
I see, now it makes sense to me. Thank you!